libraries/rushell/src/Tokenizer.ts

```diff
@@ -0,0 +1,244 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import { TextRange } from './TextRange';
+import { ParseError } from './ParseError';
+
+export enum TokenKind {
+  // One or more spaces/tabs
+  Spaces,
+  // A single newline sequence such as CRLF or LF
+  NewLine,
+  // An unrecognized character
+  Other,
+  // A sequence of characters that doesn't contain any symbols with special meaning
+  // Characters can be escaped, in which case the Token.text may differ from the
+  // Token.range.toString()
+  Text,
+
+  // The "&&" operator, which executes the following command only if the preceding command
+  // succeeded (i.e. returned a zero exit code).
+  AndIf,
+
+  // A double-quoted string which can do variable expansions
+  DoubleQuotedText,
+
+  // A dollar sign followed by an environment variable name
+  DollarVariable,
+
+  // The end of the input string
+  EndOfInput
+}
+
+export class Token {
+  public readonly kind: TokenKind;
+  public readonly range: TextRange;
+
+  /**
+   * The extracted content, which depends on the type:
+   *
+   * Text: The unescaped content
+   * DoubleQuotedText: The unescaped contents inside the quotes.
+   * DollarVariable: The variable name without the "$"
+   */
+  public readonly text: string;
+
+  public constructor(kind: TokenKind, range: TextRange, text?: string) {
+    this.kind = kind;
+    this.range = range;
+    this.text = text === undefined ? this.range.toString() : text;
+  }
+
+  public toString(): string {
+    return this.text;
+  }
+}
+
+const textCharacterRegExp: RegExp = /[a-z0-9_\\]/i;
+const startVariableCharacterRegExp: RegExp = /[a-z_]/i;
+const variableCharacterRegExp: RegExp = /[a-z0-9_]/i;
+
+export class Tokenizer {
+  public readonly input: TextRange;
+  private _currentIndex: number;
+
+  private static _isSpace(c: string | undefined): boolean {
```

Comment:
>_isSpace [](start = 17, length = 8)

It'd be interesting to find out what happens in a POSIX shell when you use a special Unicode space character.

Reply:
I tested it empirically like this:

`echo $(echo -e a '\u0020' b '\t' c)`  --> prints `"a b c"` with 1 column between each letter because %20 and '\t' are whitespace

`echo $(echo -e a '\u2002' b '\t' c)`  --> prints `"a  b c"` with 3 columns between a and b because the Unicode "EN SPACE" is treated as a normal character

Based on these experiments, my code here seems to be correct.


---
In reply to: [197314051](https://github.com/Microsoft/web-build-tools/pull/715#discussion_r197314051) [](ancestors = 197314051)